crates/types/src/hardware.rs

```diff
@@ -0,0 +1,160 @@
+use std::{
+    fmt::{Debug, Error, Formatter},
+    mem::{size_of, ManuallyDrop},
+    path::Path,
+    sync::Arc,
+    time::SystemTime,
+};
+
+use color_eyre::Result;
+use image::RgbImage;
+use spl_network_messages::{GameControllerReturnMessage, GameControllerStateMessage, SplMessage};
+
+use crate::{Rgb, YCbCr422, YCbCr444};
+
+use super::{CameraPosition, Joints, Leds, SensorData};
+
+pub trait Interface {
+    fn read_from_microphones(&self) -> Result<Samples>;
+
+    fn get_now(&self) -> SystemTime;
+    fn get_ids(&self) -> Ids;
+    fn read_from_sensors(&self) -> Result<SensorData>;
+    fn write_to_actuators(&self, positions: Joints, stiffnesses: Joints, leds: Leds) -> Result<()>;
+
+    fn read_from_network(&self) -> Result<IncomingMessage>;
+    fn write_to_network(&self, message: OutgoingMessage) -> Result<()>;
+
+    fn read_from_camera(&self, camera_position: CameraPosition) -> Result<Image>;
+}
+
+#[derive(Clone, Debug)]
+pub struct Ids {
+    pub body_id: String,
+    pub head_id: String,
+}
+
+#[derive(Clone, Default)]
+pub struct Image {
+    pub buffer: Arc<Vec<YCbCr422>>,
+    pub width_422: u32,
+    pub height: u32,
+}
```

Comment:
I'd argue the image type is too big and should be moved to a separate file/module.

Reply:
Extracted to an external file.